code/modules/food_and_drinks/kitchen_machinery/food_replicator.dm

```diff
@@ -0,0 +1,631 @@
+	var/obj/item/weapon/reagent_containers/food/snacks/loaded = null
+
+/obj/machinery/food_replicator/New()
+	..()
+	component_parts = list()
+	component_parts += new /obj/item/weapon/circuitboard/food_replicator(null)
+	component_parts += new /obj/item/weapon/stock_parts/micro_laser(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module(null)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin(null)
+	component_parts += new /obj/item/weapon/stock_parts/console_screen(null)
+	component_parts += new /obj/item/stack/cable_coil(null, 5)
+	RefreshParts()
+
+/obj/machinery/food_replicator/upgraded/New()
+	..()
+	component_parts = list()
+	component_parts += new /obj/item/weapon/circuitboard/food_replicator(null)
+	component_parts += new /obj/item/weapon/stock_parts/micro_laser/quadultra(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator/femto(null)
+	component_parts += new /obj/item/weapon/stock_parts/scanning_module/triphasic(null)
+	component_parts += new /obj/item/weapon/stock_parts/matter_bin/bluespace(null)
+	component_parts += new /obj/item/weapon/stock_parts/console_screen(null)
+	component_parts += new /obj/item/stack/cable_coil(null, 5)
+	RefreshParts()
+
+/obj/machinery/food_replicator/Destroy()
+	clear_memory(-1)		//just in case, probably don't need to do this.
+	eject_dish()
+	if(food_paste)			//if we have food paste left, let's just dump a pile of goop on the floor to represent it being wasted
+		if(food_paste >= 100)	//a lot of food paste, better make it a big pile of goop
+			new /obj/effect/decal/cleanable/molten_object/large(get_turf(src))
+		else
+			new /obj/effect/decal/cleanable/molten_object(get_turf(src))
+	return ..()
+
+/obj/machinery/food_replicator/RefreshParts()
+	recycle_rate = 0.7
+	reagent_tier = 1
+	memory_slots = 1
+	for(var/obj/item/weapon/stock_parts/manipulator/M in component_parts)
+		recycle_rate = 0.6 + (M.rating * 0.1)	//70% at tier 1, 80% at tier 2, 90% at tier 3, 100% at tier 4
+	for(var/obj/item/weapon/stock_parts/micro_laser/ML in component_parts)
+		reagent_tier = ML.rating
+	for(var/obj/item/weapon/stock_parts/scanning_module/SM in component_parts)
+		memory_slots = SM.rating
+	for(var/obj/item/weapon/stock_parts/matter_bin/MB in component_parts)
+		max_food_paste = MB.rating * 200
+
+//scan_food is used to determine the cost of replicating the loaded food item.
+//if the recycling var is set, it includes reagents from all tiers, even if you can't replicate them yet. Non-tiered reagents still don't count though.
+//this is so don't get double shafted by trying to recycle complex dishes in a low tier replicator (the recycle_rate already punishes you enough).
+/obj/machinery/food_replicator/proc/scan_food(recycling = 0)
+	var/result = 0
+	if(!loaded)
+		return -1
+	if(!loaded.reagents)
+		visible_message("<span class='warning'>[loaded] lacks a reagents holder! Alert a coder!</span>")
```

Comment:
`log_runtime(EXCEPTION("[src] lacks a reagents holder"))` or something like that would be a good idea